Render example notebooks in Doxygen documentation

nilsvu opened the pull request:
## Proposed changes

Convert notebook files in `docs/` to markdown using `jupyter-nbconvert` (this is very fast, try it for yourself). The markdown files get picked up by Doxygen automatically and show up on https://spectre-code.org:

![Bildschirmfoto 2022-02-25 um 17 31 55](https://user-images.githubusercontent.com/746230/155751785-8d9a1555-2fa5-4599-9e00-8768af44fb25.png)

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Install `nbconvert` in container

nilsvu committed:
Support markdown math in Doxygen

Use standard $...$ syntax for inline math and
\begin{equation} ... \end{equation} for display math
in markdown files. A Doxygen filter formats the math for
Doxygen, i.e., uses \f$...\f$ for inline math and
\f${equation}{ ... \f} for display math.

nilsvu committed:
Render Jupyter notebooks in Doxygen documentation

All .ipynb notebooks in docs/ get converted to markdown and
then picked up by Doxygen.

nilsvu committed:
Render BBH ID example in Doxygen

- Give the page a title that shows up in the sidebar
- List as subpage of a new "Examples" section
- Gitignore markdown output

nilsvu committed:
Use matplotlib style to render plots with white background

This fixes a display bug where axis labels are not visible on dark
backgrounds.

nilsvu commented:
CI will fail until I push the container with `jupyter-nbconvert` installed.

wthrowe commented:
Since CI can't produce an artifact, can you upload the rendered docs somewhere so we can review them?

jupyter-nbconvert should be listed as an (optional) dependency.

nilsvu committed:
Add nbconvert to optional dependencies

Can be installed with `pip install nbconvert`.

nilsvu commented:
Here's a zip:
[html.zip](https://github.com/sxs-collaboration/spectre/files/8142981/html.zip)

nilsvu commented:
Here's a zip:
[Uploading html.zip…]()

wthrowe commented on docs/Examples/BbhInitialData/.gitignore:
```diff
@@ -1,5 +1,7 @@
 # See LICENSE.txt for details.
 
+BbhInitialData.md
+BbhInitialData_files/
```
Is it possible to avoid writing to the source tree?

nilsvu replied:
I think yes. I'll give it a try.

wthrowe commented on docs/config/markdown-math-filter.sed:
```diff
@@ -0,0 +1,25 @@
+s|\\end{\([^}]*\)}\$\$|\\f}|g
+
+# Wrap inline math in \f$...\f$
+s|\([^\\f]\)\$\([^$]*\)\$|\1\\f\$\2\\f\$|g
```
This is too aggressive.  From the "Build System" guide:
```
add_spectre_library(\f${LIBRARY} \f${LIBRARY_SOURCES})
```
Similar things happen in a few other places.

I don't have any ideas for how to fix this without support in the markdown renderer.

nilsvu replied:
Good catch. How about we don't apply the math rendering in code blocks (fenced with backticks)? That's probably how markdown renderers handle this as well.

wthrowe replied:
Yes, that should do it.

wthrowe commented on docs/config/markdown-math-filter.sed:
```diff
@@ -0,0 +1,25 @@
+s|\\end{\([^}]*\)}\$\$|\\f}|g
+
+# Wrap inline math in \f$...\f$
+s|\([^\\f]\)\$\([^$]*\)\$|\1\\f\$\2\\f\$|g
```
What is excluding `\$` from the match for?  That doesn't appear anywhere in our markdown source.

nilsvu replied:
That makes the regex "lazy", so it won't gobble up `$` characters until it hits the last one on the line.

wthrowe replied:
Sorry, I don't follow.

nilsvu replied:
switched to perl, so the syntax here changed (all of this is way easier in perl)

wthrowe replied:
Unfortunately, perl scripts are banned from SpECTRE.  Because python is better or some such nonsense.

nilsvu replied:
Hm I'm not sure Python can be as fast as perl here, haven't tried though..

nilsvu replied:
Full docs generation takes 102s with no filter, 108s with Perl filter, 135s with Python filter. Thoughts?

Edit: I added this information to the code in a fixup. I should also note that I tried doing this in sed and failed.

wthrowe replied:
I'm not surprised python is slow.  The interpreter takes tens of milliseconds to start, about an order of magnitude slower than perl.

The anti-perl faction is over there -> @nilsdeppe @kidder

nilsdeppe replied:
The reasoning isn't "Python is better than Perl" (this is super subjective and application dependent!) the reasoning is that more science folks (or at least students) are comfortable with Python than perl (python is actually taught, while perl isn't), as well as that having fewer languages being used in the code base makes it more accessible. 

Having said all that, we have used single-line perl regex calls because that's faster and easier than Python. I haven't looked at how long this script would be, but I'm okay with fairly short and simple perl files if the reason is performance (33s overhead on dox build is too much IMO). What I would like to avoid is the situation in SpEC where we have something like 50% of scripts in python and 50% in perl, then have them calling back and forth. This results in very difficult to understand code, regardless of what languages are being used. It looks like this sed script could just be a few one-line perl regex calls from CMake. We used to do a lot of regex-replace in perl for Charm things, if I remember correctly.

nilsvu replied:
👍 I second all that

wthrowe commented on cmake/SetupDoxygen.cmake:
```diff
@@ -1,10 +1,29 @@
+# Targets to preprocess the documentation
+# - Convert all `.ipynb` files in `docs/` to markdown, so they get picked up by
+#   Doxygen.
+get_filename_component(_PY_BIN_DIR ${Python_EXECUTABLE} DIRECTORY)
```
When I try to run this on my machine, cmake fails with
```
CMake Error at cmake/SetupDoxygen.cmake:7 (get_filename_component):
  get_filename_component called with incorrect number of arguments
Call Stack (most recent call first):
  CMakeLists.txt:53 (include)
```

nilsvu replied:
Maybe `Python_EXECUTABLE` is empty for some reason? I'll fix it.

nilsvu replied:
should be fixed now.

wthrowe replied:
Yep.  Now it works.

nilsvu commented:
@wthrowe could you check the fixups when you get the chance?

@nilsdeppe could you do the second review, or only check the Perl usage here is agreeable?

nilsdeppe commented on cmake/SetupDoxygen.cmake:
```diff
@@ -1,10 +1,35 @@
+  message(STATUS "Found jupyter-nbconvert: ${NBCONVERT}")
+  add_custom_target(
+    doc-notebooks-to-markdown
+    COMMAND find ${CMAKE_SOURCE_DIR}/docs -name "*.ipynb"
```
How does this scale with number of notebooks? I'm worried that if we run this every time we run the docs it'll slow down docs a lot. Maybe (in the future) we should have individual notebooks be added explicitly and then tracked as a dependency and only update if the file is updated?

nilsvu replied:
In my tests this didn't take any noticeable amount of time. The command runs as a separate target before `doc` or `doc-check`, so you'll see for yourself how fast it runs. If with more notebooks it turns out to slow down I think your suggestion to run it only over updated notebooks sounds good.

nilsdeppe commented on cmake/SetupDoxygen.cmake:
```diff
@@ -1,10 +1,35 @@
 
+  # For INPUT_FILTER in Doxyfile. Using Python instead of Perl here increases
+  # docs generation time by ~25%. Runtimes are 102s (no filter), 108s (Perl) and
+  # 135s (Python) at the time of writing.
```
Add Mar, 20222 (code might get moved around later and then git will be more difficult to get the right date)

nilsdeppe commented on docs/Doxyfile.in:
```diff
@@ -130,6 +132,8 @@ EXAMPLE_PATTERNS       =
 
 EXAMPLE_RECURSIVE      = NO
 
+INPUT_FILTER           = "@PERL_EXECUTABLE@ -0777 -p @PROJECT_SOURCE_DIR@/docs/config/markdown-math-filter.pl"
```
Could you add a comment why this is doing and why?

nilsdeppe commented on docs/Doxyfile.in:
```diff
@@ -90,10 +90,12 @@ CITE_BIB_FILES         = @PROJECT_SOURCE_DIR@/docs/References.bib
         @PROJECT_SOURCE_DIR@/docs/DevGuide \
         @PROJECT_SOURCE_DIR@/docs/PublicationPolicy.md \
         @PROJECT_SOURCE_DIR@/docs \
+        @CMAKE_BINARY_DIR@/docs/tmp/notebooks_md \
```
Most of these are "obvious" why they are added (we have C++ code there), but this is in the binary dir. Could you add a comment on line 90 explaining that this is where notebooks converted to Markdown are dumped?

nilsdeppe commented on docs/Examples/plots.mplstyle:
```diff
@@ -0,0 +1,5 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
```
Could you add a comment what this is used for and some guidance on how to use it? I see it's used in the notebook, but I think it would be good to document it here as well

nilsdeppe reviewed (dismissed):
Overall looks good, just some requests for additional comments in places to help the next person who looks at this :) No need for me to re-approve anything

nilsvu commented:
All done @nilsdeppe, thanks!

wthrowe commented:
Fixups look fine.

nilsvu committed:
SetupDoxygen: find Python

nilsvu commented:
Squashed and pushed the container

nilsvu committed:
Pin Numpy version in container to fix segfaults

Numpy v1.22 segfaults in Pypp tests, see issue
https://github.com/sxs-collaboration/spectre/issues/3844.

nilsvu commented:
I had to pin Numpy to v1.21 in the container to avoid segfaults with the new v1.22, see #3844.

nilsvu commented:
Restarted CI due to #3845
